engine/utils/Colour.h

```diff
@@ -9,39 +9,203 @@
 #ifndef SIEGE_ENGINE_COLOUR_H
 #define SIEGE_ENGINE_COLOUR_H
 
+#include <cstdint>
+#include <type_traits>
+
+#include "Hash.h"
+
 namespace Siege
 {
-struct Colour
-{
-    // Public constants
 
-    static const Colour Black;
+/**
+ * The base Colour template struct
+ * @tparam T the data type used to represent RGBA values
+ * @warn this type can only have numerical template types
+ */
+template<typename T, typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+struct Colour;
 
-    static const Colour White;
+/**
+ * The float Colour type
+ */
+typedef Colour<float> FColour;
 
-    static const Colour Red;
+/**
+ * The 8-bit unsigned int Colour type. This type is recommended for general usage.
+ */
+typedef Colour<uint8_t> IColour;
+
+/**
+ * The base Colour template struct
+ * @tparam T the data type used to represent RGBA values
+ * @warn this type can only have numerical template types
+ */
+template<typename T, typename std::enable_if<std::is_arithmetic<T>::value>::type*>
```

Comment:
I'm not sure the partial specialisation gets us much except for additional complexity. If you want to keep it for now, that's fine, but I would have just defined both of them separately as I don't think we'll have any other colour classes, just colour space functionality from here on.

Reply:
As we discussed in Discord, this seems to be the only way to define a class which can be converted between similar types whilst also allowing them to be used as `constexpr`s. We can discuss later if we can find an alternative to this.